### DISCUSSION TITLE
How to opt out of supabase-js SDK caching in nextjs 14

### DISCUSSION THREAD

**Discussion by hnykda** [supabase/supabase#20022] [Questions]
Hi. I simply want to stop caching of supabase-js in my nextjs app. To prevent caching now, I add a random string to the request which is terrible, e.g.:

```
  const { data, error } = await supabase
    .from('table')
    .select('id')
    .eq('id', someId)
    // This stupidity is here just because of caching
    // for some reason, supabase underlying calls are being cached
    // unless we add this random string to the query...
    .neq('id', nanoid())
    .maybeSingle();
```
which works, but oh boy...

I wanted to turn off caching by providing a custom fetch function that would have `cache: "no-cache"` as hinted [here](https://github.com/supabase/supabase-js/issues/725#issuecomment-1578811299) (and [docs](https://supabase.com/docs/reference/javascript/initializing)), but I cannot get it work typewise, as for:

```
export const createFetch =
  (options: Pick<RequestInit, 'next' | 'cache'>) =>
  (url: RequestInfo | URL, init?: RequestInit) => {
    return fetch(url, {
      ...init,
      ...options,
    });
  };

const supabase = createClient<Database>(
  NEXT_PUBLIC_SUPABASE_URL,
  SUPABASE_SERVICE_ROLE_KEY,
  {
    auth: {
      persistSession: false,
    },
    global: {
      fetch: createFetch({
        cache: 'no-store',
      }),
    },
  },
);
```

I get:

```
Type '(url: RequestInfo | URL, init?: RequestInit) => Promise<Response>' is not assignable to type '{ (input: RequestInfo | URL, init?: RequestInit | undefined): Promise<Response>; (input: RequestInfo, init?: RequestInit | undefined): Promise<...>; }'.
  Types of parameters 'url' and 'input' are incompatible.
    Type 'RequestInfo' is not assignable to type 'RequestInfo | URL'.
      Type 'Request' is not assignable to type 'RequestInfo | URL'.
        Property 'referrer' is missing in type 'import("/Users/dan/work/varuna/code/varuna/node_modules/undici-types/fetch").Request' but required in type 'Request'.ts(2322)
lib.dom.d.ts(18717, 14): 'referrer' is declared here.
(property) fetch?: {
    (input: RequestInfo | URL, init?: RequestInit | undefined): Promise<Response>;
    (input: RequestInfo, init?: RequestInit | undefined): Promise<...>;
} | undefined
```

and also every supabase call start to complain about many other stuff I can't parse very well, like this:
![image](https://github.com/supabase/supabase/assets/2741256/cde6e354-efa7-4646-b8d0-83953672f8b4)

What can I do?

**Comment by GaryAustin1** [comment]
Are you sure the custom fetch is the best way?   
https://github.com/orgs/supabase/discussions/19084
Mentions 2 ways users here have used.  I don't know anything about next.js though and this is really a next.js question and not supabase, unless you have to do the custom fetch I guess.

**Reply by hnykda** [reply]
I don't see how I could, and the recommended things from that discussion do not work for me at all. Consider my simple setup here:

```
  const testRequests = async () => {
    const id = 'someId';
    console.log('first', await getForecastByIdAction(id));
    await new Promise((resolve) => setTimeout(resolve, 1000));
    await createForecast('somethihng', id);
    await new Promise((resolve) => setTimeout(resolve, 1000));
    console.log('second', await getForecastByIdAction(id));
    await new Promise((resolve) => setTimeout(resolve, 1000));
    console.log('third', await getForecastByIdAction(id));
  };

  useEffect(() => {
    testRequests();
  }, []);
```

where the `getForecastById` or `createForecast` are functions such as:

```
export async function getForecastById(id: ForecastId) {
  const { data, error } = await supabase
    .from('forecasts')
    .select('*')
    .eq('id', id)
    .maybeSingle();
  if (error) {
    console.error('in getForecastById: ', error);
    throw error.message;
  }
  return data as ExtendedForecast | null;
}
```

I can refresh a page that is using this component how many times I want, and it will always print just:

```
first null
second null
third null
```

until I kill the dev server and then start again (so it flushes the server fetch cache). I would expect the second call to return the forecast. It does not + it caches it even in between hard refresh (I assume the cache is process-scoped, so in distributed envs like vercel it behaves randomly depending on what machine processes that request)!

I tried adding

```
export const dynamic = 'force-dynamic'; // no caching
```

to both the (client) component as well as the (server) `page.tsx`, no luck, same behavior (I guess it's because the supabase client is fetched in entirely different modules?)

It makes me crazy, spend hours on chasing bugs because of that.


I don't see how I could get around that in Nextjs, so I wanted to find a way how I can adjust the fetch in supabase. Therefore I rather asked here. Is there anything bad about using custom fetch?

**Reply by GaryAustin1** [reply]
No worries asking here.  Hopefully someone has info for you.
This is a related discussion from your link.
https://github.com/supabase/supabase-js/issues/917

**Reply by KarthikeyanVijayanEasyLLM** [reply]
I am facing the same issue. And I use createServerClient with 'use server' for server actions in Next 14. 

If I use .neq("id", randomInt(1000000000)), I am not facing this problem

**Reply by GaryAustin1** [reply]
A user in Discord suggests this:
```js
export const supabase = createClientComponentClient({
  options: {
    global: {
      fetch: createFetch({
        cache: "no-store",
      }),
    },
  },
});

'use server'
export const createFetch =
  (options: Pick<RequestInit, "next" | "cache">) =>
  (url: RequestInfo | URL, init?: RequestInit) => {
    return fetch(url, {
      ...init,
      ...options,
    });
  };
  ```

**Reply by hnykda** [reply]
That does not work type wise for me on Next 14.0.1:

```
export const createFetch =
  (options: Pick<RequestInit, 'next' | 'cache'>) =>
  (url: RequestInfo | URL, init?: RequestInit) => {
    return fetch(url, {
      ...init,
      ...options,
    });
  };

export const supabase = createClient<Database>(
  NEXT_PUBLIC_SUPABASE_URL,
  SUPABASE_SERVICE_ROLE_KEY,
  {
    auth: {
      persistSession: false,
    },
    global: {
      fetch: createFetch({
        cache: 'no-store',
      }),
    },
  },
);
```

gives me:

```
Type '(url: RequestInfo | URL, init?: RequestInit) => Promise<Response>' is not assignable to type '{ (input: RequestInfo | URL, init?: RequestInit | undefined): Promise<Response>; (input: RequestInfo, init?: RequestInit | undefined): Promise<...>; }'.
  Types of parameters 'url' and 'input' are incompatible.
    Type 'RequestInfo' is not assignable to type 'RequestInfo | URL'.
      Type 'Request' is not assignable to type 'RequestInfo | URL'.
        Property 'referrer' is missing in type 'import("/Users/dan/work/varuna/code/varuna/node_modules/undici-types/fetch").Request' but required in type 'Request'.ts(2322)
lib.dom.d.ts(18717, 14): 'referrer' is declared here.
```

and all infered typing gets wrong, so I lose all typesafety

**Reply by danielfoxp2** [reply]
I discovered today that I have this same problem. The `.neq("id", uuidv4())` is helping as a workaround but it would be awesome to have a real solution for it.

**Comment by chaitanya-000** [comment]
Hey @hnykda, have you found some solution to this problem ? Facing the same caching issue.

**Reply by hnykda** [reply]
Nope. 😭

**Reply by chaitanya-000** [reply]
So @hnykda, about your workaround of adding the string, is it working well in production ? Asking so that I can use and implement your workaround  and move on, rather than solving this issue at the end of the sprint

**Reply by hnykda** [reply]
yes, it works fine. you need to generate valid input for that `.neq(field, ...)`

**Reply by chaitanya-000** [reply]
hey @hnykda , was looking into this whole caching thing and had some really useful insights related to the  images, are you facing issue with regular data or with images only ?

**Reply by hnykda** [reply]
Hey. Regular data as well.

**Comment by albertocubeddu** [comment]
I've tried the experimental feature and it's working for me:

```typescript
import { unstable_noStore as noStore } from "next/cache";
export function createADMINClient() {
   noStore();
   return createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.SUPABASE_SERVICE_ROLE_KEY!,
      {
         auth: {
            autoRefreshToken: false,
            persistSession: false,
         },
      }
   );
}
```

Now every query using this Client is getting fresh results.

**Reply by Larry-Osakwe** [reply]
This worked for me

Slight example for anyone else in the future

```
import { createClient } from '@supabase/supabase-js'
import { unstable_noStore as noStore } from "next/cache";

const supabaseUrl = process.env.SUPABASE_URL as string
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY as string

export function createADMINClient() {
    noStore();
    return createClient(
        supabaseUrl,
        supabaseKey,
        {
            auth: {
                autoRefreshToken: false,
                persistSession: false,
            },
        }
    );
}

export const supabase = createADMINClient();

export async function getAllEvents() {
    const { data, error } = await supabase
        .from('events')
        .select()

    if (error) {
        console.error('Error fetching events:', error);
    } else {
        console.log('Fetched events:', data.length);
    }

    return data;
}
```

**Reply by ordinz** [reply]
I'm sad to say this isn't breaking the cache for me.

**Reply by MJUrian-Learner** [reply]
Works for me

**Reply by Abdelmoghit-IDH** [reply]
This work for me too! Thank you so much bro

**Comment by tadeegan** [comment]
Where is this caching behavior documented? I can't find any docs on it. Is it a supabase thing or next thing?

**Reply by hnykda** [reply]
It's Next's [feature](https://nextjs.org/docs/app/api-reference/functions/fetch) (I had to resist here not to put it in quotes):

> Next.js extends the native [Web fetch() API](https://developer.mozilla.org/docs/Web/API/Fetch_API) to allow each request on the server to set its own persistent caching semantics.

I flipped out when I found out. I mean, I am not blaming anyone, but I would rather live in a world where this wasn't the case or it was somewhere clearly written, in BIG RED SIGN AT THE BEGINNING OF THE DOCS. Like, the lived experience is "A database select is being cached?!"

**Reply by tadeegan** [reply]
`Next.js will default to force-cache` wtf...

**Comment by horlah** [comment]
@hnykda I found a solution, for a server component simply add this:
```js
export const revalidate = 60;
// value can be 0 or any figure in seconds, it is the revalidation interval.
```
See docs reference: https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate

Or try this solution for client components https://github.com/orgs/supabase/discussions/20022#discussioncomment-9347672

**Reply by hnykda** [reply]
I can't explain why because I haven't found out, but that doesn't work for us for some reason. The app is semi-complex, importing the client from other files, using actions, ... And putting `revalidate = 0` to the `page.tsx` with the component where we experienced that didn't change anything.

Also, I would really just want to never cache the stuff with the supabase client, while keeping other cache stuff in. This solution doesn't do that, it turns of caching globally.

**Comment by sos0** [comment]
@hnykda Would it make sense to add a middleware solution that intercepts supabase domains?

**Reply by hnykda** [reply]
Eh, maybe? 🤔 How would you actually do that though?

**Comment by lmoj** [comment]
I just do this
```
const supabase = createClient(
  <SUPABASE_URL>,
  <SUPABASE_KEY>,
  {
    global: {
      fetch: (url: any, options = {}) => {
        return fetch(url, { ...options, cache: 'no-store' });
      }
    }
  }
)
```

**Reply by hnykda** [reply]
This seems to work for me typewise. Will check if I still see issues with it.

**Reply by dustinjior** [reply]
you are a hero, thank you very much for this <3

**Reply by kaifirschau** [reply]
This also worked for me! Thank you so much!

**Reply by foxbit19** [reply]
Thank you! This fixed the issue!

**Reply by rizbud** [reply]
thank you!
you saved my life ❤️

**Reply by rkefi** [reply]
Happy i've found this only after 3 hours of debugging 😅😅😅
Thank you so much!!!

**Reply by filippobrigati** [reply]
So with this, also server components are re-fetching on every reloading right?

**Reply by featherJ** [reply]
Thank you so, so much!

**Reply by Kzoeps** [reply]
Thank youuuu sooo muchhh!!!!!

**Reply by darul75** [reply]
Okey yes best workaround and working solution I found, caused me hours of debugging to figure out which option to use either in a route in NextJS or here from the client itself. Having switched to Next 15 or another project had forgotten this one was in 14, and I recommend migrating, this cache mechanism caused us so many troubles. Thank you all!

**Comment by sos0** [comment]
Next 15 is removing its default caching 😅

**Reply by denisrudov** [reply]
it is. but now I would like to opt in cache for particular request. is it possible?
